File: packages/contracts/src/local-notifications.ts
import { z } from 'zod';
import { utf8ByteLength } from './text.ts';

/**
 * The local notification contract.
 *
 * Local notifications are scheduled and presented entirely by the operating
 * system on one device. Nothing here reaches a server, so this contract exists
 * to keep the four Device SDKs describing the same behavior rather than to
 * serialize a request.
 *
 * The portable surface is deliberately small. Anything expressible on only one
 * platform belongs in `platform`, so that reading the portable fields tells you
 * what will actually happen everywhere.
 */

/**
 * Reserved identifier namespace.
 *
 * A local notification and a Cloud push both end up as an OS notification with
 * an identifier. If those identifiers shared a space, cancelling a local
 * reminder could remove a delivered Cloud notification, and an open could be
 * attributed to the wrong one. SDKs therefore prefix caller IDs with this
 * namespace before handing them to the platform, and reject caller IDs that
 * already start with it so the namespace cannot be forged.
 */
export const LOCAL_NOTIFICATION_ID_NAMESPACE = 'notifie.local.';

export const MAX_LOCAL_NOTIFICATION_ID_LENGTH = 64;
export const MAX_LOCAL_CUSTOM_DATA_KEYS = 20;
export const MAX_LOCAL_CUSTOM_DATA_BYTES = 4096;

/**
 * Identifier charset.
 *
 * Restricted because these strings cross three type systems: an iOS request
 * identifier (string), an Android notification ID (int) and a `PendingIntent`
 * request code (int). The Android integers are derived by hashing this string,
 * so it must be stable, printable and free of the whitespace and separators
 * that intent extras and log lines handle inconsistently.
 */
const localNotificationIdSchema = z
  .string()
  .min(1)
  .max(MAX_LOCAL_NOTIFICATION_ID_LENGTH)
  .regex(
    /^[A-Za-z0-9._:-]+$/,
    'Use only letters, digits, dot, underscore, colon or hyphen.',
  )
  .refine(
    (id) => !id.startsWith(LOCAL_NOTIFICATION_ID_NAMESPACE),
    `Identifiers must not start with the reserved "${LOCAL_NOTIFICATION_ID_NAMESPACE}" namespace.`,
  );

const localCustomDataSchema = z
  .record(z.string().min(1).max(128), z.string().max(1024))
  .refine(
    (data) => Object.keys(data).length <= MAX_LOCAL_CUSTOM_DATA_KEYS,
    `At most ${MAX_LOCAL_CUSTOM_DATA_KEYS} custom data fields are allowed.`,
  )
  .refine(
    (data) => utf8ByteLength(JSON.stringify(data)) <= MAX_LOCAL_CUSTOM_DATA_BYTES,
    'Custom data must be at most 4 KB.',
  )
  .refine(
    (data) => Object.keys(data).every((key) => !key.startsWith('gk_')),
    'Custom data must not use the reserved gk_ prefix.',
  );

/** Fires once at an absolute instant. */
const atScheduleSchema = z.object({
  type: z.literal('at'),
  /**
   * An instant, not a wall-clock time. If the device changes timezone before
   * it fires, this still fires at the same moment worldwide.
   */
  timestamp: z.string().datetime({ offset: true }).or(z.string().datetime()),
});

/** Fires once after a delay measured from scheduling. */
const inScheduleSchema = z.object({
  type: z.literal('in'),
  seconds: z
    .number()
    .int()
    .min(1)
    // A year. Longer intervals are better expressed as an absolute instant,
    // and both platforms become unreliable across that span anyway.
    .max(365 * 24 * 60 * 60),
});

const hourSchema = z.number().int().min(0).max(23);
const minuteSchema = z.number().int().min(0).max(59);

/**
 * Fires every day at a wall-clock time.
 *
 * Wall clock rather than a fixed interval: a 9am reminder must stay at 9am
 * across a daylight-saving transition. Repeating every 86400 seconds would
 * silently drift to 8am or 10am and stay there.
 */
const dailyScheduleSchema = z.object({
  type: z.literal('daily'),
  hour: hourSchema,
  minute: minuteSchema,
});

/** Fires every week at a wall-clock time. Monday is 1, Sunday is 7 (ISO-8601). */
const weeklyScheduleSchema = z.object({
  type: z.literal('weekly'),
  weekday: z.number().int().min(1).max(7),
  hour: hourSchema,
  minute: minuteSchema,
});

/**
 * Monthly recurrence is deliberately absent.
 *
 * "The 31st of every month" has no correct answer in February, and every
 * available behavior — skip, clamp, roll forward — silently surprises somebody.
 * An explicit sequence of `at` schedules expresses the intent without guessing.
 */
export const localScheduleSchema = z.discriminatedUnion('type', [
  atScheduleSchema,
  inScheduleSchema,
  dailyScheduleSchema,
  weeklyScheduleSchema,
]);

/**
 * Platform-specific options.
 *
 * Isolated so the portable fields stay portable. A value here affects one
 * platform and is ignored elsewhere, which is a property callers can rely on
 * rather than discover.
 */
const platformOptionsSchema = z
  .object({
    ios: z
      .object({
        /** `UNNotificationContent.threadIdentifier`, used for grouping. */
        threadId: z.string().min(1).max(64).optional(),
        /** A `UNNotificationCategory` registered by the host application. */
        categoryId: z.string().min(1).max(64).optional(),
        /** Application badge to apply when presented. */
        badge: z.number().int().min(0).max(99_999).optional(),
        /** `false` uses no sound; a string names a bundled sound file. */
        sound: z.union([z.boolean(), z.string().min(1).max(64)]).optional(),
        interruptionLevel: z
          .enum(['passive', 'active', 'timeSensitive', 'critical'])
          .optional(),
      })
      .strict()
      .optional(),
    android: z
      .object({
        /** Existing notification channel. Falls back to the SDK default. */
        channelId: z.string().min(1).max(64).optional(),
        /**
         * Requests exact delivery.
         *
         * Off by default because exact alarms require a user-visible
         * permission on Android 12+ and are a scarce system resource. The
         * scheduling result reports the precision actually granted, so a
         * denied request degrades to inexact rather than failing.
         */
        exact: z.boolean().optional(),
        /** Survives Doze. Use only for user-critical alarms such as alarms clocks. */
        allowWhileIdle: z.boolean().optional(),
        smallIconName: z.string().min(1).max(64).optional(),
        /** Grouping key, equivalent in spirit to iOS `threadId`. */
        groupKey: z.string().min(1).max(64).optional(),
      })
      .strict()
      .optional(),
  })
  .strict();

export const localNotificationSchema = z
  .object({
    /**
     * Stable, caller-owned identity. Scheduling the same ID replaces the
     * pending notification rather than adding a second one, which is what
     * makes scheduling safe to retry.
     */
    id: localNotificationIdSchema,
    title: z.string().min(1).max(100),
    body: z.string().min(1).max(250),
    schedule: localScheduleSchema,
    /** Delivered to the open handler; never opened by the SDK itself. */
    deepLink: z.string().url().optional(),
    /**
     * String-only, matching the remote contract. Platform payloads are string
     * dictionaries, so allowing richer values here would mean each SDK
     * inventing its own encoding and disagreeing at the boundary.
     */
    customData: localCustomDataSchema.optional(),
    platform: platformOptionsSchema.optional(),
  })
  .strict();

/** Why a local notification could not be scheduled. */
export const localScheduleErrorSchema = z.enum([
  /** Content or schedule failed validation. */
  'invalid_request',
  /** The user has not granted notification permission. */
  'permission_denied',
  /** The requested time has already passed. */
  'schedule_in_past',
  /**
   * The platform pending-notification limit is reached.
   *
   * iOS keeps only the 64 soonest pending requests and discards the rest in
   * silence. Reporting capacity explicitly is the difference between a
   * developer learning this now and learning it from a user.
   */
  'capacity_exceeded',
  /** The platform rejected the request for a reason the SDK cannot classify. */
  'platform_error',
]);

/**
 * Delivery precision actually granted.
 *
 * Returned rather than assumed: Android downgrades exact alarms when the
 * permission is absent, and a scheduler that claimed exactness it did not
 * receive would be lying at the moment accuracy matters.
 */
export const localSchedulePrecisionSchema = z.enum(['exact', 'inexact']);

export type LocalSchedule = z.infer<typeof localScheduleSchema>;
export type LocalNotification = z.infer<typeof localNotificationSchema>;
export type LocalNotificationInput = z.input<typeof localNotificationSchema>;
export type LocalScheduleError = z.infer<typeof localScheduleErrorSchema>;
export type LocalSchedulePrecision = z.infer<typeof localSchedulePrecisionSchema>;
export type LocalPlatformOptions = z.infer<typeof platformOptionsSchema>;

/** Namespaces a caller identifier for the platform notification store. */
export function toPlatformNotificationId(id: string): string {
  return `${LOCAL_NOTIFICATION_ID_NAMESPACE}${id}`;
}

/** Recovers a caller identifier, or null when the ID is not a Notifie local one. */
export function fromPlatformNotificationId(platformId: string): string | null {
  return platformId.startsWith(LOCAL_NOTIFICATION_ID_NAMESPACE)
    ? platformId.slice(LOCAL_NOTIFICATION_ID_NAMESPACE.length)
    : null;
}

/**
 * Resolves the next firing instant for a schedule.
 *
 * Shared so the SDKs and their tests agree on the boundary cases rather than
 * each re-deriving them: a daily time that has already passed today belongs to
 * tomorrow, and a weekly slot occurring later today must not skip a week.
 *
 * Returns null for a one-shot schedule already in the past, which callers
 * report as `schedule_in_past` rather than scheduling something that can never
 * fire.
 */
export function nextLocalOccurrence(schedule: LocalSchedule, from: Date): Date | null {
  switch (schedule.type) {
    case 'at': {
      const at = new Date(schedule.timestamp);
      return at.getTime() > from.getTime() ? at : null;
    }
    case 'in': {
      return new Date(from.getTime() + schedule.seconds * 1000);
    }
    case 'daily': {
      const next = withTime(from, schedule.hour, schedule.minute);
      if (next.getTime() <= from.getTime()) next.setDate(next.getDate() + 1);
      return next;
    }
    case 'weekly': {
      const next = withTime(from, schedule.hour, schedule.minute);
      // getDay() is 0-6 with Sunday first; the contract is ISO 1-7 with Monday
      // first, so Sunday maps to 7 rather than 0.
      const currentWeekday = next.getDay() === 0 ? 7 : next.getDay();
      let delta = schedule.weekday - currentWeekday;
      if (delta < 0) delta += 7;
      if (delta === 0 && next.getTime() <= from.getTime()) delta = 7;
      next.setDate(next.getDate() + delta);
      return next;
    }
  }
}

/**
 * Builds a local wall-clock time on the same calendar day.
 *
 * Mutating a copy rather than constructing from components keeps this in the
 * host's timezone, which is the whole point of a calendar schedule.
 */
function withTime(from: Date, hour: number, minute: number): Date {
  const next = new Date(from.getTime());
  next.setHours(hour, minute, 0, 0);
  return next;
}

File: sdks/flutter/lib/src/local_notifications.dart
/// Local notifications.
///
/// These are scheduled and presented by the operating system on this device.
/// They need no API key, no network and no Notifie account.
///
/// This file defines the portable model and validates it before anything
/// crosses the method channel. Validating in Dart means a mistake is reported
/// with a useful message on the calling thread, rather than surfacing as an
/// opaque platform exception two layers down.
///
/// The scheduling itself is never implemented here: it delegates to the native
/// Swift and Android implementations, which own persistence, reboot recovery
/// and the platform trigger APIs. A second scheduling engine would drift from
/// the first, and the subtle parts — wall-clock recurrence, capacity limits —
/// are exactly where the drift would hide.
library;

import 'dart:convert';

/// Reserved identifier namespace.
///
/// Local notifications and Cloud pushes both become OS notifications with an
/// identifier. If those shared a space, cancelling a reminder could remove a
/// delivered Cloud notification and an open could be attributed to the wrong
/// one. Caller ids are prefixed natively, and ids claiming the prefix are
/// rejected so it cannot be forged.
const String kLocalNotificationIdNamespace = 'notifie.local.';

const int kMaxLocalNotificationIdLength = 64;
const int kMaxLocalNotificationTitleLength = 100;
const int kMaxLocalNotificationBodyLength = 250;
const int kMaxLocalCustomDataKeys = 20;
const int kMaxLocalCustomDataBytes = 4096;

final RegExp _idPattern = RegExp(r'^[A-Za-z0-9._:-]+$');

/// When a local notification fires.
sealed class LocalSchedule {
  const LocalSchedule();

  Map<String, Object?> toJson();
}

/// Fires once at an absolute instant, regardless of later timezone changes.
final class LocalScheduleAt extends LocalSchedule {
  const LocalScheduleAt(this.timestamp);

  final DateTime timestamp;

  @override
  Map<String, Object?> toJson() => {
        'type': 'at',
        // ISO-8601 with an explicit offset: a bare local timestamp would be
        // reinterpreted by whichever platform parsed it.
        'timestamp': timestamp.toUtc().toIso8601String(),
      };

  @override
  bool operator ==(Object other) =>
      other is LocalScheduleAt && other.timestamp == timestamp;

  @override
  int get hashCode => timestamp.hashCode;
}

/// Fires once after an interval measured from scheduling.
final class LocalScheduleAfter extends LocalSchedule {
  const LocalScheduleAfter(this.duration);

  final Duration duration;

  @override
  Map<String, Object?> toJson() => {
        'type': 'after',
        'seconds': duration.inSeconds,
      };

  @override
  bool operator ==(Object other) =>
      other is LocalScheduleAfter && other.duration == duration;

  @override
  int get hashCode => duration.hashCode;
}

/// Fires every day at a wall-clock time.
///
/// Wall clock rather than a fixed 24-hour interval: a 9am reminder must stay at
/// 9am across a daylight-saving change rather than drifting an hour and staying
/// there.
final class LocalScheduleDaily extends LocalSchedule {
  const LocalScheduleDaily({required this.hour, required this.minute});

  final int hour;
  final int minute;

  @override
  Map<String, Object?> toJson() => {
        'type': 'daily',
        'hour': hour,
        'minute': minute,
      };

  @override
  bool operator ==(Object other) =>
      other is LocalScheduleDaily &&
      other.hour == hour &&
      other.minute == minute;

  @override
  int get hashCode => Object.hash(hour, minute);
}

/// Fires every week at a wall-clock time.
///
/// [weekday] follows ISO-8601 and `DateTime.weekday`: Monday is 1, Sunday is 7.
final class LocalScheduleWeekly extends LocalSchedule {
  const LocalScheduleWeekly({
    required this.weekday,
    required this.hour,
    required this.minute,
  });

  final int weekday;
  final int hour;
  final int minute;

  @override
  Map<String, Object?> toJson() => {
        'type': 'weekly',
        'weekday': weekday,
        'hour': hour,
        'minute': minute,
      };

  @override
  bool operator ==(Object other) =>
      other is LocalScheduleWeekly &&
      other.weekday == weekday &&
      other.hour == hour &&
      other.minute == minute;

  @override
  int get hashCode => Object.hash(weekday, hour, minute);
}

/// iOS-specific options, ignored on Android.
final class LocalNotificationIosOptions {
  const LocalNotificationIosOptions({
    this.threadId,
    this.categoryId,
    this.badge,
    this.sound,
  });

  final String? threadId;
  final String? categoryId;
  final int? badge;

  /// `null` uses the default sound, `false` is silent, a name plays a bundled
  /// file.
  final Object? sound;

  Map<String, Object?> toJson() => {
        if (threadId != null) 'threadId': threadId,
        if (categoryId != null) 'categoryId': categoryId,
        if (badge != null) 'badge': badge,
        if (sound != null) 'sound': sound,
      };
}

/// Android-specific options, ignored on iOS.
final class LocalNotificationAndroidOptions {
  const LocalNotificationAndroidOptions({
    this.channelId,
    this.exact = false,
    this.allowWhileIdle = false,
    this.groupKey,
  });

  final String? channelId;

  /// Requests exact delivery.
  ///
  /// Off by default: exact alarms need a user-visible permission on Android
  /// 12+ and are a scarce system resource. A denied request degrades to inexact
  /// and is reported through [LocalScheduleResult.precision] rather than
  /// failing.
  final bool exact;

  /// Fires during Doze. Reserve for user-critical alarms.
  final bool allowWhileIdle;

  final String? groupKey;

  Map<String, Object?> toJson() => {
        if (channelId != null) 'channelId': channelId,
        'exact': exact,
        'allowWhileIdle': allowWhileIdle,
        if (groupKey != null) 'groupKey': groupKey,
      };
}

/// A local notification to schedule.
final class LocalNotification {
  const LocalNotification({
    required this.id,
    required this.title,
    required this.body,
    required this.schedule,
    this.deepLink,
    this.customData = const {},
    this.ios,
    this.android,
  });

  /// Caller-owned, stable identity. Scheduling the same id replaces the pending
  /// notification rather than adding a second one, which is what makes
  /// scheduling safe to retry.
  final String id;
  final String title;
  final String body;
  final LocalSchedule schedule;

  /// Delivered to the notification-open handler. The SDK never opens it.
  final String? deepLink;

  /// String-only, matching the remote contract. Platform payloads are string
  /// dictionaries, so richer values would mean each SDK inventing an encoding.
  final Map<String, String> customData;

  final LocalNotificationIosOptions? ios;
  final LocalNotificationAndroidOptions? android;

  Map<String, Object?> toJson() => {
        'id': id,
        'title': title,
        'body': body,
        'schedule': schedule.toJson(),
        if (deepLink != null) 'deepLink': deepLink,
        'customData': customData,
        if (ios != null) 'ios': ios!.toJson(),
        if (android != null) 'android': android!.toJson(),
      };

  /// Returns why this notification is invalid, or null when it is acceptable.
  ///
  /// Mirrors `packages/contracts` and the native implementations so every
  /// Notifie SDK rejects the same input for the same reason.
  String? validate() {
    if (id.isEmpty || id.length > kMaxLocalNotificationIdLength) {
      return 'id must be 1-$kMaxLocalNotificationIdLength characters';
    }
    if (id.startsWith(kLocalNotificationIdNamespace)) {
      return 'id must not start with the reserved '
          '"$kLocalNotificationIdNamespace" namespace';
    }
    if (!_idPattern.hasMatch(id)) {
      return 'id may contain only letters, digits, dot, underscore, colon '
          'or hyphen';
    }
    if (title.isEmpty || title.length > kMaxLocalNotificationTitleLength) {
      return 'title must be 1-$kMaxLocalNotificationTitleLength characters';
    }
    if (body.isEmpty || body.length > kMaxLocalNotificationBodyLength) {
      return 'body must be 1-$kMaxLocalNotificationBodyLength characters';
    }
    if (customData.length > kMaxLocalCustomDataKeys) {
      return 'at most $kMaxLocalCustomDataKeys custom data fields';
    }
    if (customData.keys.any((key) => key.startsWith('gk_'))) {
      return 'custom data must not use the reserved gk_ prefix';
    }
    // Measured in UTF-8 because the platform payload budget is bytes. A
    // character count passes for text the platform then refuses to store,
    // which emoji and non-Latin scripts hit first.
    if (utf8.encode(jsonEncode(customData)).length > kMaxLocalCustomDataBytes) {
      return 'custom data must be at most 4 KB';
    }

    final schedule = this.schedule;
    switch (schedule) {
      case LocalScheduleAfter(:final duration):
        if (duration.inSeconds < 1) {
          return 'interval must be at least 1 second';
        }
      case LocalScheduleDaily(:final hour, :final minute):
        return _timeError(hour, minute);
      case LocalScheduleWeekly(:final weekday, :final hour, :final minute):
        if (weekday < 1 || weekday > 7) {
          return 'weekday must be 1-7 with Monday as 1';
        }
        return _timeError(hour, minute);
      case LocalScheduleAt():
        break;
    }
    return null;
  }

  static String? _timeError(int hour, int minute) {
    if (hour < 0 || hour > 23) return 'hour must be 0-23';
    if (minute < 0 || minute > 59) return 'minute must be 0-59';
    return null;
  }
}

/// Why a local notification could not be scheduled.
///
/// Stable across platforms so host code branches on the cause rather than
/// parsing a message. Each case needs a different response: a denied permission
/// is a product problem, a full queue is a scheduling-strategy problem, and an
/// invalid request is a programming error.
enum LocalScheduleError {
  invalidRequest,
  permissionDenied,
  scheduleInPast,

  /// iOS keeps only the 64 soonest pending notifications and silently discards
  /// the rest.
  capacityExceeded,

  platformError;

  static LocalScheduleError fromCode(String code) => switch (code) {
        'invalid_request' => LocalScheduleError.invalidRequest,
        'permission_denied' => LocalScheduleError.permissionDenied,
        'schedule_in_past' => LocalScheduleError.scheduleInPast,
        'capacity_exceeded' => LocalScheduleError.capacityExceeded,
        _ => LocalScheduleError.platformError,
      };
}

/// Delivery precision actually granted, which may be less than requested.
enum LocalSchedulePrecision {
  exact,
  inexact;

  static LocalSchedulePrecision fromCode(String? code) =>
      code == 'exact' ? LocalSchedulePrecision.exact : LocalSchedulePrecision.inexact;
}

/// The outcome of a scheduling attempt.
sealed class LocalScheduleResult {
  const LocalScheduleResult();
}

final class LocalScheduleSuccess extends LocalScheduleResult {
  const LocalScheduleSuccess({
    required this.precision,
    this.nextTrigger,
  });

  /// What the platform actually granted. Never assume the request succeeded as
  /// asked: Android downgrades exact alarms without the 12+ permission.
  final LocalSchedulePrecision precision;

  /// Absent when the platform owns the next date for a recurring schedule.
  final DateTime? nextTrigger;
}

final class LocalScheduleFailure extends LocalScheduleResult {
  const LocalScheduleFailure(this.error, [this.message]);

  final LocalScheduleError error;
  final String? message;

  @override
  String toString() => 'LocalScheduleFailure(${error.name}, $message)';
}

/// A scheduled notification awaiting delivery.
final class PendingLocalNotification {
  const PendingLocalNotification({required this.id, this.nextTrigger});

  final String id;
  final DateTime? nextTrigger;
}

/// What this device can actually do.
///
/// Queryable so an application can adapt before scheduling, rather than
/// discovering a platform limit from a user who never received a reminder.
final class LocalNotificationCapabilities {
  const LocalNotificationCapabilities({
    required this.permission,
    required this.canScheduleExactAlarms,
    required this.supportedSchedules,
    this.pendingCapacity,
  });

  final LocalNotificationPermission permission;

  /// Android 12+ gates exact alarms behind a user-visible permission. Always
  /// true on iOS, where scheduling precision is not separately permissioned.
  final bool canScheduleExactAlarms;

  final Set<String> supportedSchedules;

  /// Maximum pending notifications, where the platform defines one. iOS keeps
  /// 64; Android has no fixed documented limit, so this is null there.
  final int? pendingCapacity;

  static LocalNotificationCapabilities fromJson(Map<Object?, Object?> json) =>
      LocalNotificationCapabilities(
        permission: LocalNotificationPermission.fromCode(
          json['permission'] as String?,
        ),
        canScheduleExactAlarms: json['canScheduleExactAlarms'] as bool? ?? false,
        supportedSchedules: ((json['supportedSchedules'] as List<Object?>?) ?? const [])
            .map((value) => value.toString())
            .toSet(),
        pendingCapacity: json['pendingCapacity'] as int?,
      );
}

/// Notification permission state.
enum LocalNotificationPermission {
  /// Never asked. The prompt has not been shown, so asking can still succeed.
  notDetermined,
  granted,

  /// Refused. On iOS the prompt will not be shown again, so this is effectively
  /// permanent and the application must degrade rather than retry.
  denied,
  provisional;

  static LocalNotificationPermission fromCode(String? code) => switch (code) {
        'granted' => LocalNotificationPermission.granted,
        'denied' => LocalNotificationPermission.denied,
        'provisional' => LocalNotificationPermission.provisional,
        _ => LocalNotificationPermission.notDetermined,
      };
}

/// Resolves the next firing instant, or null for a one-shot already past.
///
/// Kept alongside the contract so Dart callers can preview a schedule without a
/// platform round trip, and so the boundary cases are asserted in Dart tests
/// against the same expectations as the TypeScript and native suites.
///
/// The native implementations remain authoritative for actual delivery.
DateTime? nextLocalOccurrence(LocalSchedule schedule, DateTime from) {
  switch (schedule) {
    case LocalScheduleAt(:final timestamp):
      return timestamp.isAfter(from) ? timestamp : null;
    case LocalScheduleAfter(:final duration):
      return from.add(duration);
    case LocalScheduleDaily(:final hour, :final minute):
      var next = DateTime(from.year, from.month, from.day, hour, minute);
      if (!next.isAfter(from)) {
        // Constructing the next calendar day rather than adding 24 hours: a
        // fixed interval drifts by an hour at a daylight-saving boundary and
        // stays there.
        next = DateTime(from.year, from.month, from.day + 1, hour, minute);
      }
      return next;
    case LocalScheduleWeekly(:final weekday, :final hour, :final minute):
      final candidate = DateTime(from.year, from.month, from.day, hour, minute);
      var delta = weekday - candidate.weekday;
      if (delta < 0) delta += 7;
      if (delta == 0 && !candidate.isAfter(from)) delta = 7;
      return DateTime(from.year, from.month, from.day + delta, hour, minute);
  }
}
